File: zero-server/src/main/kotlin/com/zeroday/handler/handlers/ResearchHandler.kt
package com.zeroday.handler.handlers

import com.zeroday.handler.HandlerContext
import com.zeroday.handler.HandlerResult
import com.zeroday.handler.MessageHandler
import com.zeroday.handler.reqStr
import com.zeroday.protocol.ErrorCategory
import com.zeroday.protocol.RequestTypes
import com.zeroday.protocol.ResponseTypes
import com.zeroday.protocol.ServerError
import kotlinx.serialization.json.JsonObject

class ResearchHandler : MessageHandler {
    override val handledTypes: Set<String> = setOf(
        RequestTypes.RESEARCH_RECIPES,
        RequestTypes.RESEARCH_INVENTORY,
        RequestTypes.RESEARCH_GATHER,
        RequestTypes.RESEARCH_START,
        RequestTypes.RESEARCH_CLAIM,
        RequestTypes.RESEARCH_USE_ITEM,
        RequestTypes.RESEARCH_STATUS
    )

    override suspend fun handle(ctx: HandlerContext, type: String, payload: JsonObject): HandlerResult {
        val playerId = ctx.authenticatedPlayerId()
            ?: return HandlerResult.Err(ServerError.fromMessage("Authentication required", ErrorCategory.AUTHENTICATION))
        return when (type) {
            RequestTypes.RESEARCH_RECIPES -> recipes(ctx, playerId)
            RequestTypes.RESEARCH_INVENTORY -> inventory(ctx, playerId)
            RequestTypes.RESEARCH_GATHER -> gather(ctx, playerId)
            RequestTypes.RESEARCH_START -> start(ctx, playerId, payload)
            RequestTypes.RESEARCH_CLAIM -> claim(ctx, playerId, payload)
            RequestTypes.RESEARCH_USE_ITEM -> useItem(ctx, playerId, payload)
            RequestTypes.RESEARCH_STATUS -> status(ctx, playerId)
            else -> HandlerResult.Err(ServerError.fromMessage("Unsupported research request: $type", ErrorCategory.VALIDATION))
        }
    }

    private suspend fun recipes(ctx: HandlerContext, playerId: String): HandlerResult {
        val recipes = ctx.services.researchService.getRecipes(playerId)
        return HandlerResult.Ok(ResponseTypes.RESEARCH_RECIPES, mapOf("recipes" to recipes))
    }

    private suspend fun inventory(ctx: HandlerContext, playerId: String): HandlerResult =
        ctx.services.researchService.getInventory(playerId).fold(
            onSuccess = { items -> HandlerResult.Ok(ResponseTypes.RESEARCH_INVENTORY, mapOf("inventory" to items)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Inventory unavailable", ErrorCategory.INTERNAL)) }
        )

    private suspend fun gather(ctx: HandlerContext, playerId: String): HandlerResult =
        ctx.services.researchService.gatherFragment(playerId).fold(
            onSuccess = { item -> HandlerResult.Ok(ResponseTypes.FRAGMENT_GATHERED, mapOf("item" to item)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Gather failed", ErrorCategory.INTERNAL)) }
        )

    private suspend fun start(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val recipeId = try { payload.reqStr("recipe_id") } catch (_: Exception) {
            return HandlerResult.Err(ServerError.fromMessage("Recipe ID required", ErrorCategory.VALIDATION))
        }
        return ctx.services.researchService.startResearch(playerId, recipeId).fold(
            onSuccess = { progress -> HandlerResult.Ok(ResponseTypes.RESEARCH_STARTED, mapOf("progress" to progress)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Research start failed", ErrorCategory.INTERNAL)) }
        )
    }

    private suspend fun claim(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val recipeId = try { payload.reqStr("recipe_id") } catch (_: Exception) {
            return HandlerResult.Err(ServerError.fromMessage("Recipe ID required", ErrorCategory.VALIDATION))
        }
        return ctx.services.researchService.claimResearch(playerId, recipeId).fold(
            onSuccess = { item -> HandlerResult.Ok(ResponseTypes.RESEARCH_CLAIMED, mapOf("item" to item)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Research claim failed", ErrorCategory.INTERNAL)) }
        )
    }

    private suspend fun useItem(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val itemId = try { payload.reqStr("item_id") } catch (_: Exception) {
            return HandlerResult.Err(ServerError.fromMessage("Item ID required", ErrorCategory.VALIDATION))
        }
        return ctx.services.researchService.useItem(playerId, itemId).fold(
            onSuccess = { msg -> HandlerResult.Ok(ResponseTypes.ITEM_USED, mapOf("message" to msg)) },
            onFailure = { error -> HandlerResult.Err(ServerError.fromMessage(error.message ?: "Use failed", ErrorCategory.INTERNAL)) }
        )
    }

    private suspend fun status(ctx: HandlerContext, playerId: String): HandlerResult {
        val status = ctx.services.researchService.getResearchStatus(playerId)
        return HandlerResult.Ok(ResponseTypes.RESEARCH_STATUS, mapOf("active_research" to status))
    }
}

File: zero-server-csharp/src/ZeroDayMMO.Server/Handlers/ResearchHandler.cs
using System.Text.Json;
using ZeroDayMMO.Server.Services;
using ZeroDayMMO.Shared;

namespace ZeroDayMMO.Server.Handlers;

public class ResearchHandler : IHandler
{
    private readonly ResearchService _researchService;
    private readonly IPlayerService _playerService;

    public string MessageType => "research_recipes";

    public ResearchHandler(ResearchService researchService, IPlayerService playerService)
    {
        _researchService = researchService;
        _playerService = playerService;
    }

    public async Task<IActionResult> HandleAsync(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });

        try
        {
            var recipes = await _researchService.GetRecipes(playerId);
            return new MessageResult("research_recipes", connectionId, new { recipes });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }

    public async Task<IActionResult> HandleInventory(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });

        try
        {
            var items = await _researchService.GetInventory(playerId);
            return new MessageResult("research_inventory", connectionId, new { inventory = items });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }

    public async Task<IActionResult> HandleGather(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });

        try
        {
            var item = await _researchService.GatherFragment(playerId);
            return new MessageResult("fragment_gathered", connectionId, new { item });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }

    public async Task<IActionResult> HandleStart(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        var recipeId = payload?.Str("recipe_id");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });
        if (string.IsNullOrEmpty(recipeId))
            return new MessageResult("error", connectionId, new { message = "Recipe ID required" });

        try
        {
            var progress = await _researchService.StartResearch(playerId, recipeId);
            return new MessageResult("research_started", connectionId, new { progress });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }

    public async Task<IActionResult> HandleClaim(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        var recipeId = payload?.Str("recipe_id");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });
        if (string.IsNullOrEmpty(recipeId))
            return new MessageResult("error", connectionId, new { message = "Recipe ID required" });

        try
        {
            var item = await _researchService.ClaimResearch(playerId, recipeId);
            return new MessageResult("research_claimed", connectionId, new { item });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }

    public async Task<IActionResult> HandleUseItem(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        var itemId = payload?.Str("item_id");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });
        if (string.IsNullOrEmpty(itemId))
            return new MessageResult("error", connectionId, new { message = "Item ID required" });

        try
        {
            var message = await _researchService.UseItem(playerId, itemId);
            return new MessageResult("item_used", connectionId, new { message });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }

    public async Task<IActionResult> HandleStatus(string connectionId, JsonElement? payload, string? requestId)
    {
        var playerId = payload?.Str("playerId");
        if (string.IsNullOrEmpty(playerId))
            return new MessageResult("error", connectionId, new { message = "playerId required" });

        try
        {
            var status = await _researchService.GetResearchStatus(playerId);
            return new MessageResult("research_status", connectionId, new { active_research = status });
        }
        catch (Exception ex)
        {
            return new MessageResult("error", connectionId, new { message = ex.Message });
        }
    }
}
